zcash_client_sqlite/src/lib.rs

```diff
@@ -1083,18 +1091,26 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                         }
                     }
                     TransferType::Incoming => {
-                        match spending_account_id {
-                            Some(id) => {
-                                if id != *output.account() {
-                                    panic!("Unable to determine a unique account identifier for z->t spend.");
-                                }
-                            }
-                            None => {
-                                spending_account_id = Some(*output.account());
-                            }
-                        }
-
                         wallet::sapling::put_received_note(wdb.conn.0, output, tx_ref, None)?;
+
+                        if let Some(account_id) = funding_account {
+                            // Even if the recipient address is external, record the send as internal.
```

Comment:
Blocking:

This is the exact misuse of the `sent_notes` account column that I mentioned earlier. The enum is named `WalletInternal`, but prior to now AFAICT this has always been used as "account-internal". Given that _any_ account in the wallet can be the funding account here, "wallet-internal" is ill-defined; the two accounts may have no common controller of spend authority (e.g. not be derived from the same seed). For example, it would make no sense for a note sent from a wallet account to ZecPages to be considered a "wallet-internal" transaction just because the ZecPages full viewing key is imported.

As an additional datapoint for this column currently meaning "account-internal", if we detect a note as going to an external address, and then later call `create_account` such that the external address is learned to actually be a wallet address, we do not update the `sent_notes` table to correct the information.

If we are going to make this change to how funding accounts are tracked, we need to do so in a way that ensures data consistency. That means either writing a migration to distinguish "account-internal" from "wallet-internal" in the `sent_notes` table, or writing a migration to redefine the `to_account_id` column as "amongst the accounts derived within or imported into this wallet database" and add code to rescan the _entire_ wallet to update relationships whenever a new account is created or imported.

Reply:
The easy fix here is to restrict this check to verify that the funding account is the receiving account. When I refer to "external" here I'm simply meant the zip32 external scope.

Reply:
Wait a minute though, this is just populating the "from_account" column of the sent_notes table correctly; the to_account column should already be getting the correct data. So I don't think there is a problem here at all.

Reply:
The recipient value being constructed in this fashion ensures that the to_account field gets set correctly. You're correct though that reusing the "WalletInternal" variant in this fashion is a change in semantics. It may be that we just need an additional Recipient variant, that can carry both the address (for external addresses) and the recipient account ID.